exchanges/bithumb/bithumb_websocket.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		panic(err)
+	}
+}
+
+// WsConnect initiates a websocket connection
+func (b *Bithumb) WsConnect() error {
+	if !b.Websocket.IsEnabled() || !b.IsEnabled() {
+		return errors.New(stream.WebsocketNotEnabled)
+	}
+
+	var dialer websocket.Dialer
+	dialer.HandshakeTimeout = b.Config.HTTPTimeout
+	dialer.Proxy = http.ProxyFromEnvironment
+
+	err := b.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return fmt.Errorf("%v - Unable to connect to Websocket. Error: %w",
+			b.Name,
+			err)
+	}
+	go b.wsReadData()
+	b.setupOrderbookManager()
+	return nil
+}
+
+// wsReadData receives and passes on websocket messages for processing
+func (b *Bithumb) wsReadData() {
+	b.Websocket.Wg.Add(1)
+	defer b.Websocket.Wg.Done()
+
+	for {
+		resp := b.Websocket.Conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		err := b.wsHandleData(resp.Raw)
+		if err != nil {
+			b.Websocket.DataHandler <- err
+		}
+	}
+}
+
+func (b *Bithumb) wsHandleData(respRaw []byte) error {
+	var resp WsReponse
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
+	}
+
+	if len(resp.Status) > 0 {
+		if resp.Status == "0000" {
+			return nil
+		}
+		return fmt.Errorf("%s: %w", resp.ResponseMessage, errWsSubFailure)
+	}
+
+	switch resp.Type {
+	case "ticker":
+		var tick WsTicker
+		err = json.Unmarshal(resp.Content, &tick)
+		if err != nil {
+			return err
+		}
+		var lu time.Time
+		lu, err = time.ParseInLocation(tickerTimeLayout,
+			tick.Date+tick.Time,
+			location)
+		if err != nil {
+			return err
+		}
+		b.Websocket.DataHandler <- &ticker.Price{
+			ExchangeName: b.Name,
+			AssetType:    asset.Spot,
+			Last:         tick.PreviousClosePrice,
+			Pair:         tick.Symbol,
+			Open:         tick.OpenPrice,
+			Close:        tick.ClosePrice,
+			Low:          tick.LowPrice,
+			High:         tick.HighPrice,
+			QuoteVolume:  tick.Value,
+			Volume:       tick.Volume,
+			LastUpdated:  lu,
+		}
+	case "transaction":
+		if !b.IsSaveTradeDataEnabled() {
+			return nil
+		}
+
+		var trades WsTransactions
+		err = json.Unmarshal(resp.Content, &trades)
+		if err != nil {
+			return err
+		}
+
+		var lu time.Time
+		for x := range trades.List {
+			lu, err = time.ParseInLocation(tradeTimeLayout,
+				trades.List[x].ContractTime,
+				location)
+			if err != nil {
+				return err
+			}
+
+			err = b.AddTradesToBuffer(trade.Data{
+				Exchange:     b.Name,
+				AssetType:    asset.Spot,
+				CurrencyPair: trades.List[x].Symbol,
+				Timestamp:    lu,
+				Price:        trades.List[x].ContractPrice,
+				Amount:       trades.List[x].ContractAmount,
+			})
+			if err != nil {
+				return err
+			}
+		}
+	case "orderbookdepth":
+		var orderbooks WsOrderbooks
+		err = json.Unmarshal(resp.Content, &orderbooks)
+		if err != nil {
+			return err
+		}
+		init, err := b.UpdateLocalBuffer(&orderbooks)
+		if err != nil && !init {
+			return fmt.Errorf("%v - UpdateLocalCache error: %s", b.Name, err)
+		}
```

Comment:
Just to clarify, you want to dismiss all errors while `init` is `true`?

Reply:
Yeah I was following what was going on in Binance. I am thinking of consolidating a lot of this functionality into the stream/buffer package.